### PR TITLE
[MDETR] Phrase grounding eval

### PR DESCRIPTION
**Pull request by ebsmothers**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #200
* #199
* #198
* __->__ #181
* #180
* #179

Add remaining components and main loop for phrase grounding evaluation.

Tested via:
```
CUBLAS_WORKSPACE_CONFIG=:4096:8 python -m torch.distributed.launch --nproc_per_node=2 --use_env phrase_grounding.py
--resume /data/home/ebs/data/mdetr/pretrained_resnet101_checkpoint.pth\?download\=1 --ema --eval
--dataset_config /data/home/ebs/torchmultimodal/examples/mdetr/phrase_grounding.json
```

Results:
Averaged stats:
+-------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+
|   Recall@k  |        all         |      animals       |     bodyparts      |      clothing      |    instruments     |       other        |       people       |       scene        |      vehicles      |
+-------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+
|   Recall@1  | 0.8244994110718492 | 0.9292543021032504 | 0.6414048059149723 | 0.8822269807280514 | 0.8129032258064516 | 0.6894031668696712 | 0.890934113194564  | 0.7912589693411611 | 0.863905325443787  |
|   Recall@5  | 0.9289821935841475 | 0.9732313575525813 | 0.8243992606284658 | 0.9588865096359743 | 0.9419354838709677 | 0.8608404384896468 | 0.9664544985377602 | 0.9041095890410958 | 0.9349112426035503 |
|  Recall@10  | 0.9489364650453821 | 0.9789674952198852 | 0.8613678373382625 | 0.9751605995717345 | 0.9548387096774194 | 0.8955542021924482 | 0.9772922759332531 | 0.9321591650358774 | 0.9556213017751479 |
| Upper_bound | 0.9848957250744821 | 0.9980879541108987 | 0.9279112754158965 | 0.9944325481798715 | 0.967741935483871  | 0.9658952496954933 | 0.9955272664717013 | 0.9856490541422048 | 0.9970414201183432 |
+-------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+--------------------+

Differential Revision: [D38160119](https://our.internmc.facebook.com/intern/diff/D38160119)

### CODE REVIEW

File: examples/mdetr/data/flickr_eval.py [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+            flickr_path,
+            subset=subset,
+            topk=top_k,
+            iou_thresh=iou_thresh,
+            merge_boxes=merge_boxes,
+        )
+        self.predictions = []
+        self.results = None
+
+    def accumulate(self):
```
Comment (RdoubleA) [inline examples/mdetr/data/flickr_eval.py:386]: remove this method?

File: examples/mdetr/data/flickr_eval.py [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+        [
+            np_boxes[:, 0].min(),
+            np_boxes[:, 1].min(),
+            np_boxes[:, 2].max(),
+            np_boxes[:, 3].max(),
+        ]
+    ]
+
+
+class RecallTracker:
```
Comment (RdoubleA) [inline examples/mdetr/data/flickr_eval.py:178]: should this go under `utils/metric.py`?

File: examples/mdetr/utils/metrics.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+        total_time_str = str(datetime.timedelta(seconds=int(total_time)))
+        print(
+            "{} Total time: {} ({:.4f} s / it)".format(
+                header, total_time_str, total_time / len(iterable)
+            )
+        )
+
+
+@torch.no_grad()
+def accuracy(output, target, topk=(1,)):
```
Comment (RdoubleA) [inline examples/mdetr/utils/metrics.py:191]: could you replace this with `torchmetrics.Precision`? https://torchmetrics.readthedocs.io/en/stable/classification/precision.html?highlight=precision

either way, I would rename this to precision anyway

File: examples/mdetr/phrase_grounding.json [hunk 4]
```diff
@@ -0,0 +1,8 @@
+{
+    "combine_datasets": ["flickr"],
+    "combine_datasets_val": ["flickr"],
+    "GT_type" : "separate",
+    "flickr_img_path" : "",
+   "flickr_dataset_path" : "",
+   "flickr_ann_path" : ""
```
Comment (RdoubleA) [inline examples/mdetr/phrase_grounding.json:7]: are these supposed to be empty? I see `flickr_dataset_path` is used but this is not defined

Comment (ebsmothers) [reply examples/mdetr/phrase_grounding.json:7]: Yeah I am following how they do things in the MDETR repo here. The user needs to fill this in with the local path containing the relevant files. Leaving it empty as I don't want to use my own local filepaths

File: examples/mdetr/utils/metrics.py (hunk 3 above)
Comment (ebsmothers) [reply examples/mdetr/utils/metrics.py:191]: Actually removing as this is not used

**Comment by ebsmothers** [comment]
@ebsmothers has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38160119).

**Comment by ebsmothers** [comment]
@ebsmothers has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38160119).

**Comment by ebsmothers** [comment]
@ebsmothers has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38160119).

**Comment by ebsmothers** [comment]
@ebsmothers has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38160119).

**Comment by ebsmothers** [comment]
@ebsmothers has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38160119).
